services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crossregion/S3CrossRegionAsyncClient.java

```diff
@@ -15,80 +15,124 @@
 import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Function;
 import software.amazon.awssdk.annotations.SdkInternalApi;
-import software.amazon.awssdk.awscore.AwsRequestOverrideConfiguration;
-import software.amazon.awssdk.endpoints.Endpoint;
+import software.amazon.awssdk.regions.Region;
 import software.amazon.awssdk.services.s3.DelegatingS3AsyncClient;
 import software.amazon.awssdk.services.s3.S3AsyncClient;
-import software.amazon.awssdk.services.s3.endpoints.S3EndpointParams;
-import software.amazon.awssdk.services.s3.endpoints.S3EndpointProvider;
+import software.amazon.awssdk.services.s3.model.S3Exception;
 import software.amazon.awssdk.services.s3.model.S3Request;
+import software.amazon.awssdk.utils.CompletableFutureUtils;
 
 @SdkInternalApi
 public final class S3CrossRegionAsyncClient extends DelegatingS3AsyncClient {
+
+    private final Map<String, Region> bucketToRegionCache = new ConcurrentHashMap<>();
+
     public S3CrossRegionAsyncClient(S3AsyncClient s3Client) {
         super(s3Client);
     }
 
     @Override
-    protected <T extends S3Request, ReturnT> CompletableFuture<ReturnT>
-            invokeOperation(T request, Function<T, CompletableFuture<ReturnT>> operation) {
+    protected <T extends S3Request, ReturnT> CompletableFuture<ReturnT> invokeOperation(
+        T request, Function<T, CompletableFuture<ReturnT>> operation) {
 
         Optional<String> bucket = request.getValueForField("Bucket", String.class);
 
         if (!bucket.isPresent()) {
             return operation.apply(request);
         }
+        String bucketName = bucket.get();
 
-        return operation.apply(requestWithDecoratedEndpointProvider(request, bucket.get()))
-                        .whenComplete((r, t) -> handleOperationFailure(t, bucket.get()));
-    }
+        if (bucketToRegionCache.containsKey(bucketName)) {
+            return operation.apply(requestWithDecoratedEndpointProvider(request,
+                                                                        () -> bucketToRegionCache.get(bucketName),
+                                                                        serviceClientConfiguration().endpointProvider().get()));
+        }
 
-    private void handleOperationFailure(Throwable t, String bucket) {
-        //TODO: handle failure case
+        CompletableFuture<ReturnT> returnFuture = new CompletableFuture<>();
+        operation.apply(request)
+                 .whenComplete((r, t) -> {
+                     if (t != null) {
+                         if (isS3RedirectException(t.getCause())) {
+                             bucketToRegionCache.remove(bucketName);
+                             Optional<String> bucketRegionFromException =
+                                 getBucketRegionFromException((S3Exception) t.getCause());
+
+                             if (bucketRegionFromException.isPresent()) {
```

Comment:
Suggestion: extract this logic (line 67 - line 74) to a separate method to make this method smaller.